Improve Drag and Drop with Index-Based Element Selection

gaomeng1900 opened the pull request:
## Problem
Current drag and drop implementation relies on XPath/CSS selectors, which LLMs often generate incorrectly due to limited context. The LLM only receives simplified element data in format `[index]<tag >text />`, making it difficult to generate accurate selectors. This results in a low success rate for drag and drop operations.

#1981 

## Solution
- Added index-based element selection as the primary method for drag and drop, just like other actions
- Updated DOM tree building to generate indices for drag/drop elements
- Maintained existing XPath/coordinate methods as fallback (for now)
    
<!-- This is an auto-generated description by cubic. -->
---

## Summary by cubic
Improved drag and drop by switching to index-based element selection, making actions more reliable and less dependent on fragile selectors.

- **New Features**
  - Added support for selecting drag and drop elements by index.
  - Kept existing selector and coordinate methods as fallback options.

<!-- End of auto-generated description by cubic. -->

gaomeng1900 commented:
> Great fixes, thanks!
> 
> We may want to clean up the `DragDropAction` to not ask for both xpath and element index, as the LLM is likely to get confused if both xpath fields and element idx fields are present in the action?
> 
> @gaomeng1900 can you share a .gif or screenrecording of this working in action?
> 
> Then I'm down to merge it!

I agree. Maybe seperate them into different actions or remove Xpath? 

Currently the simplified HTML data is not enough to generate a currect xpath. But after several failed attempts. A model may be able to give a loose css selector like `[name="item 1"]` or `*:contains("item 1")`.

mertunsall commented:
Hi @gaomeng1900 - I left an important comment on the PR. Can we make it such that there is a primary way of doing drag and drop and not multiple? I think we can remove xpath and just use index - is there something we can do to tell the model that there is a drag and drop possibility?

Also, let's please fix the types to get the PR ready to merge.

mertunsall commented:
I just realized @pirate made the same comment - I think we can get rid of xpath and use indices instead.

gaomeng1900 commented:
I removed the xpath approach and got rid of the offsets (they didn’t make sense since the LLM doesn’t know the size of the elements). I simplified the implementation, added a test case for it, it's in a new file because test_controller.py was getting too long.

@pirate @mertunsall

gaomeng1900 commented on browser_use/controller/views.py:
```diff
@@ -67,7 +67,17 @@ class Position(BaseModel):
 
 class DragDropAction(BaseModel):
-	# Element-based approach
+	# Index-based approach (preferred)
```
done

mertunsall commented:
Hi! We decided to remove drag and drop from default actions due to high
token consumption and rare usage - can you please make an example with the
action instead? you can put it in folder examples/custom-functions

On Wed 2 Jul 2025 at 14:42, Simon ***@***.***> wrote:

> @gaomeng1900 <https://github.com/gaomeng1900> requested your review on:
> #1980 <https://github.com/browser-use/browser-use/pull/1980> Improve Drag
> and Drop Functionality with Index-Based Element Selection.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/browser-use/browser-use/pull/1980#event-18431351279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIEPRZMFK3ZN52VA5BVWR33GPHSVAVCNFSM6AAAAAB7MM2HFWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJYGQZTCMZVGEZDOOI>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

gaomeng1900 commented:
> Hi! We decided to remove drag and drop from default actions due to high token consumption and rare usage - can you please make an example with the action instead? you can put it in folder examples/custom-functions
> […](#)
> On Wed 2 Jul 2025 at 14:42, Simon ***@***.***> wrote: @gaomeng1900 <https://github.com/gaomeng1900> requested your review on: #1980 <#1980> Improve Drag and Drop Functionality with Index-Based Element Selection. — Reply to this email directly, view it on GitHub <[#1980 (comment)](https://github.com/browser-use/browser-use/pull/1980#event-18431351279)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFIEPRZMFK3ZN52VA5BVWR33GPHSVAVCNFSM6AAAAAB7MM2HFWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJYGQZTCMZVGEZDOOI> . You are receiving this because your review was requested.Message ID: ***@***.***>

I can, of course, move the controller to an example. However, it still depends on the current detection of interactive elements.
Would you like to keep the drag and drop detection in buildDomTree.js, or should we design a way to extend the HTML dehydration process?

gaomeng1900 commented:
> @gaomeng1900 we are working heavily on the agent currently, maybe wait a couple weeks and come back to this? I dont want to waste your time with many merge conflicts
> 
> we have lots of ideas internally on how to try to improve success rates here, but based on our eval data we need to reduce the total number of actions exposed to the LLM, so we're cutting anything that's buggy or less commonly needed until every default action is rock solid

I've moved the drag-and-drop feature to the examples. Maybe you can merge this?

I’m okay with handling conflicts. Focusing on the basics is the right move. Maybe make it more expandable so those quirky features on quirky websites can be handled with custom actions and hooks?
